File: c++/022.cpp
// Copyright 2012 Bob Wilkinson <bob@fourtheye.org>

// This is a solution for http://projecteuler.net/problem=22

#include <cstdio>
#include <fstream>  // NOLINT
// Fairly natural use of a stream to read file to a string?
#include <string>
#include <vector>
#include <algorithm>

int main() {
  std::string filename = "../names.txt";
  std::ifstream names_file(filename);
  if (names_file.bad()) {
    printf("ERROR: Cannot open %s\n", filename.c_str());
    return 1;
  }

  std::string name_string;
  if (names_file.is_open())
    names_file >> name_string;

  std::vector< std::string > names;
  int start = 1;
  for (unsigned int i = 1; i < name_string.length(); ++i) {
    if (name_string[i] == '"' && name_string[i+1] == ','
        && name_string[i+2] == '"') {
      names.push_back(name_string.substr(start, i - start));
      start = i + 3;
    }
  }
  names.push_back(name_string.substr(start, name_string.length() - 1));

  std::sort(names.begin(), names.end());

  int result = 0;
  for (unsigned int i = 0; i < names.size(); ++i) {
    int letters = 0;
    for (unsigned int j = 0; j < names.at(i).length(); ++j)
      letters += names.at(i).at(j) - 64;
    result += letters * (i+1);
  }

  printf("%d\n", result);
  return(0);
}

File: python/022.py
#!/usr/bin/env python3

def main():
  """ this computes projecteuler.net problem 022 """

  f = open('../names.txt', 'r')
  name_string = f.read()

  names = []
  start = 1
  for i in range(len(name_string) - 3):
    if (name_string[i] == '"'):
      if (name_string[i+1] == ','):
        if (name_string[i+2] == '"'):
          name = name_string[start:i]
          names.append(name)
          start = i + 3

  name = name_string[start:len(name_string) - 1]
  names.append(name)
  names.sort()

  total = 0
  for i in range(len(names)):
    letters = 0
    for j in range(len(names[i])):
      letters += ord(names[i][j]) - 64
    total += letters * (i+1)

  print(total)
  return 0

if __name__ == "__main__":
    main()
